### PR TITLE
Adds program record functionality

### LINKED ISSUE #1208: Add learner program records
As a learner on MITx Online, I'd like to be able to access a record of my progress through a program I'm enrolled in, similar to how I can do this in MicroMasters. 

This will be cross-porting the Program Record functionality from MicroMasters to MITx Online, including these features:
- Overview of the learner's progress through the courses in the program
- Sharing a copy of the record directly to a partner institution
- Generating a revokable link to view the record that can be shared by the learner

The first attempt at this will lean heavily on the MicroMasters look and feel (for the record-specific parts); it is expected that the design team will have a MITx Online-specific design in the future.

#### Designs and Mockups

See the Program Record in MicroMasters:
image

#### Acceptance Criteria:

- [ ] The program record is accessible through the dashboard
- [ ] The dashboard link is hidden behind a toggle (feature flag and URL param; like was done for the program UI) so it isn't viewable by the public until it has been reviewed
- [ ] The program record displays the program name and a tag indicating the overall progress the learner has made (Partially Completed or Completed)
- [ ] It displays the learner's name, username and email
- [ ] It displays each course in the program, with a tag for its status (core/required or elective) in the program, along with the course ID, grade earned, verified attempts, certificate date earned, and certificate status for the learner
- [ ] It allows record sharing as a whole to be toggled on or off
- [ ] Once record sharing is allowed, it allows records to be sent to a partner institution (which then cannot be revoked)
- [ ] Once record sharing is allowed, it allows a shareable link to be generated to allow an unauthenticated users to view the learner's record. This link should not be available before the learner activates it and should be deactivated if the learner revokes record sharing. 

#### Out of Scope

- Not sure just yet

### PR DESCRIPTION
**Pull request by jkachel**
#### Pre-Flight checklist

- [X] Screenshots and design review for any changes that affect layout or styling
  - [X] Desktop screenshots
  - [ ] Mobile width screenshots
- [X] Migrations
  - [X] Migration is backwards-compatible with current production code
- [X] Testing
  - [X] Code is tested
  - [X] Changes have been manually tested
- [X] Settings
  - [X] New settings are documented and present in `app.json`
  - [X] New settings have reasonable development defaults, if applicable
  - [X] Opened issue for DevOps regarding necessary configuration changes to deployed environments

#### What are the relevant tickets?

Fixes #1208 

#### What's this PR do?

Adds learner program records to the system:
- The learner can access their record by accessing the View Program Record link in the program drawer for the program.
- The record shows basic info about the learner and the program, and lists out the courses in the program, with the learner's status in each. 
   - The record has a tag for Partially Completed or Completed depending on their completion of the courses in the program.
   - Each course is tagged with Not Enrolled or Certificate Earned tags appropriately.
   - Each course lists the highest grade the learner achieved in the course, if there is one.
   - Each course is tagged with either Core or Elective, depending on where it sits in the requirements tree.
- The learner can activate record sharing for the record.
   - Upon activation, an anonymous share record is created. This uses a UUID to identify the particular learner record.
   - The learner can activate the Share button to retrieve a link to the anonymous share, which can then be copied and sent to other people at the learner's choosing.
   - The Share dialog has a Copy Link button which will copy the link text to the clipboard.
   - The learner can activate the Send Learner Record button, which allows the learner to send their record directly to a partner school configured within the system. 
      - Partner school shares have their own identifying UUID and exist separately from the anonymous share.
      - Partner schools are alerted to new shared learner records via email.
   - The learner can revoke sharing.
      - Revoking sharing invalidates the active anonymous record share UUID.
      - Revoking sharing _does not_ invalidate any partner records that have been shared. The records can be invalidated or deleted on the admin side, though.
      - Re-sharing the record generates a _new_ UUID for the anonymous share - _old sharing links will never work again_. 
 - Anonymous users can activate a shared link (either from a partner school email or from the learner directly).
    - Anonymous viewing displays the same information as is displayed to the learner.
    - Anonymous viewing does not provide sharing controls for the record. 
- Record sharing is hidden behind a feature flag.
   - The feature flag is `ENABLE_LEARNER_RECORDS`.
   - Additionally, the program UI needs to be turned on - the link to activate is in code that is hidden if the program UI is off.
   - As with the program UI, the learner record code can be enabled by appending `enable_lr` to the URL in some way. (The code that checks for these things checks for the code as part of the GET parameters - if you need to enable both learner records and the program UI via the URL, appending something like `?enable_programs&enable_lr` should work.) 
    
#### How should this be manually tested?

1. Ensure you have a learner account enrolled in a program.
2. Generate grades and certificates for the learner.
3. Using Django Admin, also create some partner school records. 
4. View the learner's dashboard and open the program drawer. You should see the View Program Record link.
5. Activate the link. The Program Record should open in a new tab.
6. Check to see that the courses are displayed properly, the learner's info is correct, and the tags are also correct. 
7. Enable sharing by clicking Allow Access. Once this is done, the buttons at the top should be Send Learner Record, Share, and Revoke Access.
8. Activate the Share button. You should see a link to the learner record, and you should be able to copy it using the button The button should indicate that the copy was successful.
9. In an incognito window or separate browser, navigate to the link that was copied. You should see the learner record without any sharing controls. Save the URL for this for later testing.
10. Activate the Send Learner Record button. You should be presented with a dialog to share the learner record to the schools configured in step 3. 
11. Select a school and send the record. You should receive an email with a link in it. 
12. Activate the link in a separate browser or incognito window. You should see the program record minus any sharing controls.
13. Revoke sharing by clicking Revoke Sharing. You should get a notification that the revocation was successful and the buttons should change back to the single Allow Sharing button.
14. Try the share link from step 8 again. It should fail.
15. Try the share link from the email generated in step 11 again. It should succeed. 
16. Re-activate sharing. Then, click Share. The link should be different. 
17. Reload the program record. The record should render with the full sharing controls enabled (rather than just Allow Sharing). 

#### Screenshots (if appropriate)

Initial record screen: 
image

After allowing sharing:
image

Dialog - Send Learner Record:
image

Dialog - Share:
image

Anonymous View:
image

### CODE REVIEW

**Review by annagav — CHANGES_REQUESTED** [review]
Could we add a bit of padding to the buttons, and I am not sure if there was design for that, but keep it consistent with the color of buttons:
Screen Shot 2022-11-30 at 3 29 58 PM

**Comment by annagav** [comment]
Also on the page with all 3 buttons, every time I hover over the buttons the whole records page shifts down.

**Comment by jkachel** [comment]
Comments should be resolved in https://github.com/mitodl/mitxonline/pull/1237/commits/da7fcdd7a1e081ce7b2061491e8b02dc60785431 . Couple notes:
- The button color (the green) is used in the other buttons on the dashboard modals (for example, the email settings dialog). We don't have actual styling on this, other than "looks like it does on MicroMasters", though.
- The changes include a reworking of how the program status tag (at top, under the program and learner info) is implemented. It should now respect the requirements of the program. (So if you've passed all the required courses and only have to pass one elective, and you have, you should see Completed. It wouldn't display Completed unless you completed all the courses prior to now.)

File: frontend/public/src/factories/course.js [hunk 1]
```diff
@@ -134,3 +146,183 @@ export const makeProgramEnrollment = (): ProgramEnrollment => ({
...
+): LearnerRecord => {
+  // This does a lot of things. It will generate a program with a handful of
+  // courses, some of which are required and some are electives, with the
+  // requisite requirements tree, and will generate grades for your fake learner
+  // depending on the flag.
+
+  const partnerSchools = []
+  const courses = []
+
+  for (let i = 0; i < 5; i++) {
```
Comment (annagav) [inline frontend/public/src/factories/course.js:239]: Why this is 5 here?

File: frontend/public/src/factories/course.js [hunk 2]
```diff
@@ -134,3 +146,183 @@ export const makeProgramEnrollment = (): ProgramEnrollment => ({
...
+          operator_value: 1,
+          title:          "Elective Courses"
+        },
+        children: []
+      }
+    ]
+  }
+
+  // add courses to the Required Courses node
+  for (let i = 0; i < 3; i++) {
```
Comment (annagav) [inline frontend/public/src/factories/course.js:290]: where the number 3 comes from?

**Review by annagav — CHANGES_REQUESTED** [review]
Just a few questions

File: courses/models.py [hunk 3]
```diff
@@ -886,6 +910,15 @@ def is_ended(self):
     def get_audit_class(cls):
         return CourseRunEnrollmentAudit
 
+    @property
+    def highest_grade(self):
+        """Returns the highest grade achieved for the course run."""
+        return (
+            self.grades.filter(course_run=self, user=self.user)
```
[the remark this answers is not in the document]
Comment (jkachel) [reply courses/models.py:917]: This is just for the individual course run. (It's a convenience property that I added but ultimately didn't use for this; the serializer looks at grades across all course runs.)

File: frontend/public/src/factories/course.js (hunk 1 above)
Comment (jkachel) [reply frontend/public/src/factories/course.js:239]: I wanted more than one partner school in there and 5 seemed like a nice number, no other reason

File: frontend/public/src/factories/course.js (hunk 2 above)
Comment (jkachel) [reply frontend/public/src/factories/course.js:290]: DEDP on RC has 3 required courses. Similarly, it has 2 electives and 2 nested electives.

File: courses/models.py [hunk 4]
```diff
@@ -1238,3 +1271,52 @@ class Meta:
...
+
+    def __str__(self):
+        return self.name
+
+
+class LearnerProgramRecordShare(TimestampedModel):
+    """
+    Tracks the sharing status of an individual learner's program record.
+
+    If partner_school is blank, then the
```
[the remark this answers is not in the document]
Comment (jkachel) [reply courses/models.py]: Fixed in https://github.com/mitodl/mitxonline/pull/1237/commits/11d7aba8374962a97420f240033176e4aec5076d

**Review by annagav — APPROVED** [review]
LGTM

**Pull request merged**